[SNAP-2985] Fixes for multiple inconsistency issue in snapshot

suranjan opened the pull request:
1. Were returning RegionEntry to iterator, which could have its value changed
  by different writer before iterator could consume it. Now returning a NonLocalRegionEntry if entry from region has to be returned.
  2. The NonLocalRegionEntry had oldDiskId and ColumFormatValue used to read from DiskEntry
  instead of the value in memory. Setting that to null now, which means the value in memory will
  be read for NonLocalRegionEntry
  3. Used CustomEntryConcurrentHashMap to make sure we remove entry only if the same entry is present in map
  4. Similarly, made sure that before a queue is removed, its size is 0 and anytime an entry is added to queue it is put again in map for atomicity
  5. While checking whehter an entry in region exists in the snapshot of a tx, avoided the entry that was being changed so that due to uncommitted entry a valid entry doesn't get removed from the oldEntryMap
  6. Added tests where bulk puts and bulk reads are checked for atomicity

## Changes proposed in this pull request

(Fill in the changes here)

## Patch testing

(Fill in the details about how this patch was tested)

## Is precheckin with -Pstore clean?

## ReleaseNotes changes

(Does this change require an entry in ReleaseNotes? If yes, has it been added to it?)

## Other PRs 

(Does this change require changes in other projects- snappydata, spark, spark-jobserver, aqp? Add the links of PR of the other subprojects that are related to this change)

Commit:
Fixes for multiple inconsistency issue in snapshot

  1. Were returning RegionEntry to iterator, which could have its value changed
  by different writer before iterator could consume it
  2. The NonLocalRegionEntry had oldDiskId and ColumFormatValue used to read from DiskEntry
  instead of the value in memory. Setting that to null now, which means the value in memory will
  be read
  3. Used CustomEntryConcurrentHashMap to make sure we remove entry only if the same entry is present in map
  4. Similarly, made sure that before a queue is removed, its size is 0 and anytime an entry is added to queue
  it is put again in map for atomicity
  5. While checking whehter an entry in region exists in the snapshot of a tx, avoided the entry that was being changed
  so that due to uncommitted entry a valid entry doesn't get removed from the oldEntryMap
  6. Added tests where bulk puts and bulk reads are checked for atomicity

Commit:
Changed test to sync betwen operations threads

  Instead of looping and quit early if writer finishes

Commit:
Some cleanup

Commit:
Some cleanup

Commit:
Merge branch 'snappy/master' into SNAP-2985

Commit:
Rverted a change and added log changes

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/store/GemFireContainer.java:
```diff
@@ -2521,7 +2521,7 @@ public long getRowSize() {
         final RowLocation rl = (RowLocation)itr.next();
         Assert.assertTrue(rl != null, "unexpected null encountered");
-        if (rl.isUpdateInProgress()) {
+        if (rl.isUpdateInProgress() || (rl.getValueWithoutFaultIn(this) instanceof Token)) {
```
much more efficient to use rl.isDestroyedOrRemoved()

sumwale reviewed (commented):
one quick comment; will add more comments later

Commit:
Set diskId to null in case of on-heap column value

Commit:
Use better check for perf

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/GemFireCacheImpl.java:
```diff
@@ -857,6 +870,8 @@ public void runOldEntriesCleanerThread(){
     // Keep each entry alive for at least 20 secs.
     Random r = new Random();
+    MapCallback queueRemover = new QueueRemover();
+    MapCallback entryRemover = new MapCallbackAdapter();
```
Why is entryRemover required since it is just the default MapCallbackAdapter? Why not just use the two argument remove() method which should be equivalent?

suranjan replied:
Will use it. was confused to see the condition being passed null there.

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/GemFireCacheImpl.java:
```diff
@@ -949,18 +964,24 @@ public void run() {
+
+      if (removedVal == null) return;
+
+      if (removedVal == re) {
```
This will always be true with the above remove call which will check for value to be equal to re else will not do the remove. Also two argument remove can be used above as mentioned before.

suranjan replied:
Will remove it. It was for testing purpose.

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/GemFireCacheImpl.java:
```diff
@@ -926,11 +941,11 @@ public void run() {
               if (entry.getValue() instanceof BlockingQueue) {
-                if (((BlockingQueue) entry.getValue()).isEmpty()) {
-                  regionEntryMap.remove(entry.getKey());
+                if (((BlockingQueue)entry.getValue()).isEmpty()) {
```
This isEmpty check is not required since it is handled by the queueRemover already. Also can have some rare race condition where queue becomes empty just after the check so better to do in the segment lock as taken by remove call below.

suranjan replied:
My thinking was to reduce contention a bit for cases where the same keys are being modified a lot.

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/GemFireCacheImpl.java:
```diff
@@ -949,18 +964,24 @@ public void run() {
+        if (!region.reservedTable() && region.needAccounting()) {
+          NonLocalRegionEntry nre = (NonLocalRegionEntry)re;
+          region.freePoolMemory(nre.getValueSize(), nre.isForDelete());
         }
```
The above accounting should only be done if it is a heap buffer since off-heap does it internally in its release. So set a flag after the release call above to skip this accounting for that case. This also assumes that accounting to acquire pool memory was also done and only done for heap buffers (off-heap does it during the buffer allocation itself).

suranjan replied:
We do the accounting while storing in oldEntryMap however don't check for heap or offHeap. Will remove the accounting for off-heap at both the place.

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/GemFireCacheImpl.java:
```diff
@@ -1074,15 +1096,15 @@ boolean notRequiredByAnyTx( LocalRegion region, RegionEntry re,
       for (TXStateProxy txProxy : getTxManager().getHostedTransactionsInProgress()) {
         TXState txState = txProxy.getLocalTXState();
-        if ((txState != null && !txState.isClosed() && TXState.checkEntryInSnapshot
+        if ((txState != null && !txState.isClosed() && TXState.checkEntryInSnapshotWithoutMyChange
```
What's "checkEntryInSnapshotWithoutMyChange"? Please change the name to reflect what it is supposed to be.

suranjan replied:
The check doesn't consider changes done by tx as in snapshot. This is to avoid a case where non-committed entry can cause committed entry to be removed.

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/NonLocalRegionEntry.java:
```diff
@@ -89,12 +89,18 @@ protected NonLocalRegionEntry(RegionEntry re, LocalRegion br,
         v = re.getValueInVMOrDiskWithoutFaultIn(br);
+        if (v instanceof SerializedDiskBuffer && !GemFireCacheImpl.hasNewOffHeap()) {
+          ((SerializedDiskBuffer)v).setDiskEntry(null, br);
+        }
```
add comment above as to why DiskEntry is being set to null for heap buffers

sumwale reviewed (commented):
Still to look at the new checkEntryInSnapshotWithoutMyChange method. Reviewed the rest and added some more comments.

sumwale commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/GemFireCacheImpl.java:
```diff
@@ -1095,6 +1117,26 @@ boolean notRequiredByAnyTx( LocalRegion region, RegionEntry re,
+        Object context, Object removeParams) {
+      if (value != null
+          && (value == NO_OBJECT_TOKEN || ((existingValue == value)
+          && ((BlockingQueue)existingValue).size() == 0))) {
```
linked queue size() implementation might not be efficient so always better to use isEmpty()

Commit:
Review comments

  Check isEmpty and default remove(key,value) method of
  CustomEntryConcurrentHashMap

Commit:
Don't account for SerializeDiskBuffer in off-heap

Commit:
For ObjectiStore use different getExecRow

  Corrected handling of getExec row to avoid ClassCastException
